thrust/thrust/mr/pool.h

```diff
@@ -305,23 +309,43 @@ class unsynchronized_pool_resource final
                     {
                         if (previous != &m_cached_oversized)
                         {
-                            oversized_block_descriptor previous_desc = **previous;
-                            previous_desc.next_cached = desc.next_cached;
-                            **previous = previous_desc;
+                            *previous = desc.next_cached;
                         }
                         else
                         {
                             m_cached_oversized = desc.next_cached;
                         }
 
                         desc.next_cached = oversized_block_descriptor_ptr();
+
+                        auto ret =
+                            static_cast<char_ptr>(static_cast<void_ptr>(ptr)) -
+                            desc.size;
+
+                        if (bytes != desc.size) {
+                            desc.current_size = bytes;
+
+                            ptr = static_cast<oversized_block_descriptor_ptr>(
+                                static_cast<void_ptr>(ret + bytes));
+
+                            if (detail::pointer_traits<
+                                    oversized_block_descriptor_ptr>::
+                                    get(desc.prev)) {
```

Comment:
suggestion: This would read much better if we added an alias:
```suggestion
                            using pointer_traits = detail::pointer_traits<oversized_block_descriptor_ptr>;
                            if (pointer_traits::get(desc.prev)) {
```